crates/rune-macros/src/context.rs

```diff
@@ -111,6 +115,60 @@ pub(crate) struct FieldProtocol {
     custom: Option<syn::Path>,
 }
 
+pub(crate) struct TypeProtocol {
+    protocol: syn::Ident,
+    handler: Option<syn::Path>,
+}
+
+impl TypeProtocol {
+    pub fn expand(&self) -> TokenStream {
+        if let Some(handler) = &self.handler {
+            let protocol = &self.protocol;
+            return quote_spanned! {protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::#protocol, #handler)?;
+            };
+        }
+        match self.protocol.to_string().as_str() {
+            "ADD" => quote_spanned! {self.protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::ADD, |this: Self, other: Self| this + other)?;
+            },
+            "STRING_DISPLAY" => quote_spanned! {self.protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::STRING_DISPLAY, |this: &Self, buf: &mut String| {
+                    use ::std::fmt::Write as _;
+                    ::std::write!(buf, "{this}")
+                })?;
+            },
+            "STRING_DEBUG" => quote_spanned! {self.protocol.span()=>
+                module.associated_function(rune::runtime::Protocol::STRING_DEBUG, |this: &Self, buf: &mut String| {
+                    use ::std::fmt::Write as _;
+                    ::std::write!(buf, "{this:?}")
+                })?;
+            },
+            _ => syn::Error::new_spanned(
+                &self.protocol,
+                format!(
+                    "`{}` is not a protocol supported for automatic generation on a type",
```

Comment:
Hm, what about "Rune protocol `{}` cannot be derived"?

Would be nice to make it clearer where the error message originates from, and for it to be a bit more concise.